### DISCUSSION TITLE
Lazy load from inverse side of 1-1 relation fails

### DISCUSSION THREAD

**Discussion by snigdha920** [mikro-orm/mikro-orm#5402] [General]
I have a 1-1 relation between Author and Book.
- A book has an author, and it is the owning side of the relation.
- An author can have a book, but it is not required.
- When I try to load a book from an author, I get the following error:

```typescript
@Entity()
@ObjectType()
export class Author {
  @PrimaryKey()
  _id!: ObjectId;

  @SerializedPrimaryKey()
  @Field(() => ID)
  public id!: string;

  @Property()
  @Field()
  name: string;

  @OneToOne(() => 'Book', (book: Book) => book.author, {
    wrappedReference: true,
    nullable: true,
  })
  public book?: Ref<Book>;

  constructor(name: string) {
    this.name = name;
  }
}
```

```typescript
@Entity()
@ObjectType()
export class Book {
  @PrimaryKey()
  _id!: ObjectId;

  @SerializedPrimaryKey()
  @Field(() => ID)
  public id!: string;

  @Property()
  @Field()
  name: string;

  @Field(() => Author, { nullable: true })
  @OneToOne(() => 'Author', (author: Author) => author.book, {
    wrappedReference: true,
    owner: true,
  })
  public author!: Ref<Author>;

  constructor(name: string) {
    this.name = name;
  }
}
```

```typescript
const book = await author.book.load(); // Error: Cannot read properties of undefined (reading 'load')
```

I expect it to return `null` if the author does not have a book.

I'm not sure what I'm missing? Minimum reproduction repo here: https://github.com/snigdha920/mikro-orm-1-1-relation

**Comment by B4nan** [comment]
> I expect it to return null if the author does not have a book.

If it's `null`, it's `null`, there won't be a `Reference` instance at all, `author.book` will be `null`.

>   public author!: Ref<Author>;

Here you are lying to the compiler, as `author` is in fact optional property.

Also unless you are using ts-morph, you are missing `ref: true` on both of those properties.

And one last hint, doing `@OneToOne(() => 'Book')` is not necessary, you don't need to use strings if you use callback, and you don't need to use callbacks if you use strings. So either `@OneToOne(() => Book)` or `@OneToOne('Book')`. If you use strings, you wont get typesafety, so better to use the callback signature - but with the actual class to allow inference. Then you don't need to type the second parameter explicitly: `@OneToOne(() => Book, book => book.author, { ... })`

**Comment by snigdha920** [comment]
> > I expect it to return null if the author does not have a book.
> 
> If it's `null`, it's `null`, there won't be a `Reference` instance at all, `author.book` will be `null`.

Even when the author has a book, `author.book.load()` throws `Cannot read properties of undefined (reading 'load')`, that's the issue

> 
> > public author!: Ref;
> 
> Here you are lying to the compiler, as `author` is in fact optional property.
> 

A book always has an `author`, since that definition is on the `Book` entity, that should be correct

**Reply by B4nan** [reply]
> Even when the author has a book, author.book.load() throws Cannot read properties of undefined (reading 'load'), that's the issue

It doesn't throw, `author.book` is `undefined` and you are trying to call `load` method on that, its not about the implementation.

I checked your repro and you are using mongo - mongo has no joins, so the mechanism that automatically fetches the owning ID cannot work. What you are trying wont work with mongo, you won't have the value there unless you populate it explicitly.

**Reply by snigdha920** [reply]
So any relation with mongo can only be accessed from the owning side or from the inverse side by populate, but not load or loadItems?

**Reply by B4nan** [reply]
> but not load or loadItems?

Yes (for `Reference.load`, it's not relevant for `Collection.loadItems`). Those methods require the FK value which you don't know for 1:1 inverse side. Without the value, there is no Reference. It's only a problem for 1:1 inverse side really, as other inverse sides are defined with `Collection` instance so they don't suffer from this problem.

I think it should also work with `em.populate()`:

```diff
-author.book.load()
+em.populate(author, ['book'])
```

And as long as your entity is managed (loaded from database), you can use its EM reference, which would be the one used with `Reference.load`:

```ts
wrap(author, true).__em!.populate(author, ['book'])
```
